Contribution guidelines for plot functions (#92)
pyfar/gallery: docs/contribute/contribution_guidelines.rst

In case of deprecations, it must also be tested if the deprecated feature is properly removed (see `tests/test_deprecations.py` for examples).

Testing plotting functions
~~~~~~~~~~~~~~~~~~~~~~~~~~

> The above also holds for plotting functions, however, in this case it is required that the plots generate baseline and test plots. Baseline plots are visually inspected when they are first created and serve as a reference. Test plots are created each time the test are ran. If a pull request changes a plot function, the test plots are visually compared against the baseline plots.

Reviewer: maybe add the case for new plots, like creating baselines

Author: I tried to make this more clear and now call them baseline and test images instead of plots. This should better distinguish between plot functions and the images that they produce.

> ``pyfar.testing.plot_utils`` contains the function ``create_figure`` which is best practice for creating a figure in a way that produces almost identical results across operating systems. The function ``save_and_compare`` is best practice for saving and comparing test plots automatically. However, the automatic comparison is usually disabled. It often fails because the plots generated by Matplotlib depend on the operating system. Different operating systems create offsets of a few pixels resulting in large differences between test and baseline plots.

Reviewer: it might make sence to add the pyfar.testing module to the documentation, since we need it in other packages. then we can also directly link to it here. we can also create a link in the module to this page.

Author: I agree that it would be good to document this. I think the best way to do this would be to add pyfar.testing to a separate section of the official docs and then link it here. I would suggest to do this in a separate pull.

In case of deprecations, it must also be tested if the deprecated feature is properly removed (see `tests/test_deprecations.py` for examples).

> Testing plotting functions

Reviewer 2: ```suggestion
Testing plot functions
```

~~~~~~~~~~~~~~~~~~~~~~~~~~

> The above also holds for plot functions, however, in this case it is required that the tests generate baseline and test images. The following documents the current best practice. See `test_plot.py <https://github.com/pyfar/pyfar/blob/main/tests/test_plot.py>`_ for a code example.

Reviewer 3: I don't think that linking to the file is good practice here. People who are new and are looking for the correct information have to search the right line of code in a large file. 
Please provide a code snippet with a minimal working example

> The above also holds for plotting functions, however, in this case it is required that the plots generate baseline and test images. The following documents the current best practice. See `test_plot.py <https://github.com/pyfar/pyfar/blob/main/tests/test_plot.py>`_ for a code example.

Reviewer 2: ```suggestion
The above also holds for plot functions, however, in this case it is required that the tests generate baseline and test images. The following documents the current best practice. See `test_plot.py <https://github.com/pyfar/pyfar/blob/main/tests/test_plot.py>`_ for a code example.
```

Baseline images show how an image produced by a given plot command (for example `pf.plot.time(pf.Signal[0, 1, 0], 1)`) should look like. Baseline images must be created when a new plot function or test is introduced, and they must be overwritten when the behavior of a plot function changes, e.g., because of a bug fix. Baseline images must be visually inspected as part of a pull request when they are created or changed.

> Test images show how an image produced by a given plot command currently looks like. Test images must be created each time the tests run. If the behavior of a pot function changes, the corresponding test image will change. In this case, the changed test image must be visually compared to the corresponding baseline image as part of a pull request.

Reviewer 2: ```suggestion
Test images show how an image produced by a given plot function in the testing environment looks like. Test images must be created each time the tests run. If the behavior of a plot function changes, the corresponding test image will change. In this case, the changed test image must be visually compared to the corresponding baseline image as part of a pull request.
```

> ``pyfar.testing.plot_utils`` contains the function ``create_figure`` for creating a figure in a way that produces almost identical results across operating systems. The function ``save_and_compare`` is best practice for saving and comparing test plots automatically. However, the automatic comparison is usually disabled. It often fails because the plots generated by Matplotlib depend on the operating system. Different operating systems create offsets of a few pixels resulting in large differences between test and baseline plots.

Reviewer 2: I'd suggest providing this information on the issue with automated comparison before introducing the baseline images.

Author: You mean, I should link it here: https://github.com/pyfar/spharpy/pull/96 ?

> ``pyfar.testing.plot_utils`` contains the function ``create_figure`` for creating a figure in a way that produces almost identical results across operating systems. The function ``save_and_compare`` is best practice for saving and comparing test plots automatically. However, the automatic comparison is usually disabled. It often fails because the plots generated by Matplotlib depend on the operating system. Different operating systems create offsets of a few pixels resulting in large differences between test and baseline plots.

Reviewer 2: ```suggestion
``pyfar.testing.plot_utils`` contains the function ``create_figure`` for creating a figure in a way that produces almost identical results across operating systems. The function ``save_and_compare`` is best practice for saving and comparing test plots automatically. However, the automatic comparison is usually disabled. It often fails because the plots generated by Matplotlib depend on the operating system. Different operating systems create offsets of a few pixels resulting in large differences in numerical error estimates between test and baseline plots.
```
